File: stats/stats_log_api_gen/test_java/host/src/com/android/test/statslogapigen/StatsHistogramTest.java
/*
 * Copyright (C) 2025 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.android.test.statslogapigen;

import static com.google.common.truth.Truth.assertThat;

import com.google.common.collect.ImmutableList;
import com.google.testing.junit.testparameterinjector.TestParameterInjector;
import com.google.testing.junit.testparameterinjector.TestParameter;
import com.google.testing.junit.testparameterinjector.TestParameterValuesProvider;

import android.util.statslogapigen.StatsHistogram;
import org.junit.Test;
import org.junit.runner.RunWith;

import java.util.Arrays;

@RunWith(TestParameterInjector.class)
public class StatsHistogramTest {

    private static final class StatsHistogramProvider extends TestParameterValuesProvider {
        @Override
        protected ImmutableList<?> provideValues(Context context) {
            return ImmutableList.of(
                value(StatsHistogram.createLinearBins(0f, 50f, 5)).withName("linear"),
                value(StatsHistogram.createExponentialBins(5f, 160f, 5)).withName("exponential"),
                value(StatsHistogram.createExplicitBins(-10, -7, 0, 19, 100)).withName("explicit"));
        }
    }

    @Test
    public void testLinearInitialState() throws Exception {
        StatsHistogram hist = StatsHistogram.createLinearBins(0f, 50f, 5);
        assertThat(hist.getBinCounts()).isEqualTo(new int[7]);
        assertThat(hist.getBins()).isEqualTo(
                new float[] {StatsHistogram.UNDERFLOW, 0, 10, 20, 30, 40, 50});
    }

    @Test
    public void testExponentialInitialState() throws Exception {
        StatsHistogram hist = StatsHistogram.createExponentialBins(5f, 160f, 5);
        assertThat(hist.getBinCounts()).isEqualTo(new int[7]);
        assertThat(hist.getBins()).isEqualTo(
                new float[] {StatsHistogram.UNDERFLOW, 5, 10, 20, 40, 80, 160});
    }

    @Test
    public void testExplicitInitialState() throws Exception {
        StatsHistogram hist = StatsHistogram.createExplicitBins(-10, -7, 0, 19, 100);
        assertThat(hist.getBinCounts()).isEqualTo(new int[6]);
        assertThat(hist.getBins()).isEqualTo(
                new float[] {StatsHistogram.UNDERFLOW, -10, -7, 0, 19, 100});
    }

    @Test
    public void testSingleEntryLinear() throws Exception {
        StatsHistogram hist = StatsHistogram.createLinearBins(0f, 50f, 5);
        hist.addValue(18);
        assertThat(hist.getBinCounts()).isEqualTo(new int[] {0, 0, 1, 0, 0, 0, 0});
    }

    @Test
    public void testSingleEntryExponential() throws Exception {
        StatsHistogram hist = StatsHistogram.createExponentialBins(5f, 160f, 5);
        hist.addValue(101);
        assertThat(hist.getBinCounts()).isEqualTo(new int[] {0, 0, 0, 0, 0, 1, 0});
    }

    @Test
    public void testSingleEntryExplicit() throws Exception {
        StatsHistogram hist = StatsHistogram.createExplicitBins(-10, -7, 0, 19, 100);
        hist.addValue(0);
        assertThat(hist.getBinCounts()).isEqualTo(new int[] {0, 0, 0, 1, 0, 0});
    }

    @Test
    public void testUnderflow(
            @TestParameter(valuesProvider = StatsHistogramProvider.class) StatsHistogram hist)
            throws Exception {
        hist.addValue(-100);
        int[] binCounts = hist.getBinCounts();
        assertThat(binCounts[0]).isEqualTo(1);

        // Check that all elements after first element are 0.
        assertThat(Arrays.stream(binCounts).skip(1).distinct()).containsExactly(0);
    }

    @Test
    public void testOverflow(
            @TestParameter(valuesProvider = StatsHistogramProvider.class) StatsHistogram hist)
            throws Exception {
        hist.addValue(200);
        int[] binCounts = hist.getBinCounts();
        assertThat(binCounts[binCounts.length - 1]).isEqualTo(1);

        // Check that all elements before last element are 0.
        assertThat(Arrays.stream(binCounts)
                .limit(binCounts.length - 1).distinct()).containsExactly(0);
    }

    @Test
    public void testAddUnderflowMarker(
            @TestParameter(valuesProvider = StatsHistogramProvider.class) StatsHistogram hist)
            throws Exception {
        hist.addValue(StatsHistogram.UNDERFLOW);
        int[] binCounts = hist.getBinCounts();
        assertThat(binCounts[0]).isEqualTo(1);

        // Check that all elements after first element are 0.
        assertThat(Arrays.stream(binCounts).skip(1).distinct()).containsExactly(0);
    }

    @Test
    public void testMultipleEntriesLinear() throws Exception {
        StatsHistogram hist = StatsHistogram.createLinearBins(0f, 50f, 5);
        hist.addValue(18);
        hist.addValue(40);
        hist.addValue(45);
        hist.addValue(19.99999f);
        hist.addValue(27);
        hist.addValue(0.0000001f);
        assertThat(hist.getBinCounts()).isEqualTo(new int[] {0, 1, 2, 1, 0, 2, 0});
    }

    @Test
    public void testMultipleEntriesExponential() throws Exception {
        StatsHistogram hist = StatsHistogram.createExponentialBins(5f, 160f, 5);
        hist.addValue(101);
        hist.addValue(40);
        hist.addValue(45);
        hist.addValue(159.99999f);
        hist.addValue(160.000001f);
        hist.addValue(80);
        assertThat(hist.getBinCounts()).isEqualTo(new int[] {0, 0, 0, 0, 2, 3, 1});
    }

    @Test
    public void testMultipleEntriesExplicit() throws Exception {
        StatsHistogram hist = StatsHistogram.createExplicitBins(-10, -7, 0, 19, 100);
        hist.addValue(0);
        hist.addValue(-10);
        hist.addValue(1);
        hist.addValue(25);
        hist.addValue(49);
        hist.addValue(-2);
        assertThat(hist.getBinCounts()).isEqualTo(new int[] {0, 1, 1, 2, 2, 0});
    }

    @Test
    public void testClear(
            @TestParameter(valuesProvider = StatsHistogramProvider.class) StatsHistogram hist)
            throws Exception {
        for (float v = -20; v <= 200; v += 3) {
            hist.addValue(v);
        }
        hist.clear();

        // Check that all elements are 0.
        assertThat(Arrays.stream(hist.getBinCounts()).distinct()).containsExactly(0);
    }
}

File: stats/stats_log_api_gen/test_cc/host/StatsHistogram_test.cpp
/*
 * Copyright (C) 2025 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <StatsHistogram.h>
#include <gmock/gmock.h>
#include <gtest/gtest.h>

#include <memory>

namespace android {
namespace util {
namespace statslogapigen {
namespace {

using namespace testing;
using std::string;
using std::unique_ptr;
using std::vector;

struct TestCase {
    string name;
    unique_ptr<StatsHistogram> (*histCreator)();
};

void PrintTo(const TestCase& testCase, std::ostream* os) {
    *os << testCase.name;
}

class StatsHistogramTestParameterized : public TestWithParam<TestCase> {
protected:
    unique_ptr<StatsHistogram> hist;

    void SetUp() override {
        hist = GetParam().histCreator();
    }
};

const vector<TestCase> testCases = {
        {"linear", []() { return StatsHistogram::createLinearBins(0, 50, 5); }},
        {"exponential", []() { return StatsHistogram::createExponentialBins(5, 160, 5); }},
        {"explicit", []() { return StatsHistogram::createExplicitBins({-10, -7, 0, 19, 100}); }}};

INSTANTIATE_TEST_SUITE_P(StatsHistogramTestParameterized, StatsHistogramTestParameterized,
                         ValuesIn(testCases));

TEST(StatsHistogramTest, InitialStateLinear) {
    unique_ptr<StatsHistogram> hist = StatsHistogram::createLinearBins(0, 50, 5);
    EXPECT_THAT(hist->getBins(), ElementsAre(UNDERFLOW_BIN, 0, 10, 20, 30, 40, 50));
    EXPECT_THAT(hist->getBinCounts(), ElementsAre(0, 0, 0, 0, 0, 0, 0));
}

TEST(StatsHistogramTest, InitialStateExponential) {
    unique_ptr<StatsHistogram> hist = StatsHistogram::createExponentialBins(5, 160, 5);
    EXPECT_THAT(hist->getBins(), ElementsAre(UNDERFLOW_BIN, 5, 10, 20, 40, 80, 160));
    EXPECT_THAT(hist->getBinCounts(), ElementsAre(0, 0, 0, 0, 0, 0, 0));
}

TEST(StatsHistogramTest, InitialStateExplicit) {
    unique_ptr<StatsHistogram> hist = StatsHistogram::createExplicitBins({-10, -7, 0, 19, 100});
    EXPECT_THAT(hist->getBins(), ElementsAre(UNDERFLOW_BIN, -10, -7, 0, 19, 100));
    EXPECT_THAT(hist->getBinCounts(), ElementsAre(0, 0, 0, 0, 0, 0));
}

TEST(StatsHistogramTest, SingleEntryLinear) {
    unique_ptr<StatsHistogram> hist = StatsHistogram::createLinearBins(0, 50, 5);
    hist->addValue(18);
    EXPECT_THAT(hist->getBinCounts(), ElementsAre(0, 0, 1, 0, 0, 0, 0));
}

TEST(StatsHistogramTest, SingleEntryExponential) {
    unique_ptr<StatsHistogram> hist = StatsHistogram::createExponentialBins(5, 160, 5);
    hist->addValue(101);
    EXPECT_THAT(hist->getBinCounts(), ElementsAre(0, 0, 0, 0, 0, 1, 0));
}

TEST(StatsHistogramTest, SingleEntryExplicit) {
    unique_ptr<StatsHistogram> hist = StatsHistogram::createExplicitBins({-10, -7, 0, 19, 100});
    hist->addValue(0);
    EXPECT_THAT(hist->getBinCounts(), ElementsAre(0, 0, 0, 1, 0, 0));
}

TEST_P(StatsHistogramTestParameterized, Underflow) {
    hist->addValue(-100);

    const vector<int> binCounts = hist->getBinCounts();
    EXPECT_EQ(binCounts[0], 1);

    const vector<int> remainingBins(binCounts.begin() + 1, binCounts.end());
    EXPECT_THAT(remainingBins, Each(Eq(0)));
}

TEST_P(StatsHistogramTestParameterized, Overflow) {
    hist->addValue(200);

    const vector<int> binCounts = hist->getBinCounts();
    EXPECT_EQ(binCounts.back(), 1);

    const vector<int> remainingBins(binCounts.begin(), binCounts.end() - 1);
    EXPECT_THAT(remainingBins, Each(Eq(0)));
}

TEST_P(StatsHistogramTestParameterized, UnderflowMarker) {
    hist->addValue(UNDERFLOW_BIN);

    const vector<int> binCounts = hist->getBinCounts();
    EXPECT_EQ(binCounts[0], 1);

    const vector<int> remainingBins(binCounts.begin() + 1, binCounts.end());
    EXPECT_THAT(remainingBins, Each(Eq(0)));
}

TEST(StatsHistogramTest, MultipleEntriesLinear) {
    unique_ptr<StatsHistogram> hist = StatsHistogram::createLinearBins(0, 50, 5);
    hist->addValue(18);
    hist->addValue(40);
    hist->addValue(45);
    hist->addValue(19.99999);
    hist->addValue(27);
    hist->addValue(0.0000001);
    EXPECT_THAT(hist->getBinCounts(), ElementsAre(0, 1, 2, 1, 0, 2, 0));
}

TEST(StatsHistogramTest, MultipleEntriesExponential) {
    unique_ptr<StatsHistogram> hist = StatsHistogram::createExponentialBins(5, 160, 5);
    hist->addValue(101);
    hist->addValue(40);
    hist->addValue(45);
    hist->addValue(159.99999);
    hist->addValue(160.000001);
    hist->addValue(80);
    EXPECT_THAT(hist->getBinCounts(), ElementsAre(0, 0, 0, 0, 2, 3, 1));
}

TEST(StatsHistogramTest, MultipleEntriesExplicit) {
    unique_ptr<StatsHistogram> hist = StatsHistogram::createExplicitBins({-10, -7, 0, 19, 100});
    hist->addValue(0);
    hist->addValue(-10);
    hist->addValue(1);
    hist->addValue(25);
    hist->addValue(49);
    hist->addValue(-2);
    EXPECT_THAT(hist->getBinCounts(), ElementsAre(0, 1, 1, 2, 2, 0));
}

TEST_P(StatsHistogramTestParameterized, Clear) {
    for (float v = -20; v <= 200; v += 3) {
        hist->addValue(v);
    }
    hist->clear();

    // Check that all elements are 0.
    const vector<int> binCounts = hist->getBinCounts();
    EXPECT_THAT(binCounts, Each(Eq(0)));
}

}  // namespace
}  // namespace statslogapigen
}  // namespace util
}  // namespace android
